### PR TITLE
Selfhosting

### PR DESCRIPTION
**Pull request by 3036662**
IF LIBRUM_SELFHOSTED != true - nothing will be changed.
----------------
IF  LIBRUM_SELFHOSTED == true

1. UserBlobStorage and BooksBlobStorage classes will be substituted by UserLocalStorage and BooksLocalStorage with Dependecy Injection mechanism
2. EmailSender.cs will use variable  SMTPMailFrom to fill email sender author.
3. AIService.cs will send simple string informing that OpenAi not available if no API token exists, otherwise it will act normal.
4. UserConroller will use it's own http  methods to show simple "password reset" form.
5. Program.cs will create local database structure if not exists.
PS.
BookRepository.cs string to Double parsing method was changed due problems with running on linux (it was throwing exception)
------------------------
I would like you to take a look, I'm pretty shure we can work it out to run perfect!
For now we can run this server as linux service in fully self-hosted mode using install instruction from my fork.
At the same time azure services will continue their normal work if not self-hosted.

### CODE REVIEW

**Comment by DavidLazarescu** [comment]
Thank you, this looks great. I'll look over it tomorrow and give you detailed feedback on it.

File: src/Application/Utility/EmailSender.cs [hunk 1]
```diff
@@ -47,10 +54,24 @@ public async Task SendEmailConfirmationEmail(User user, string token)
...
     {
-        var resetLink = $"https://librumreader.com/resetPassword?email={user.Email}&token={token}";
-        
+		// go to librum site if not selfhosted
+		var resetLink = $"https://librumreader.com/resetPassword?email={user.Email}&token={token}";
+		// if self hosted change resetlink
+		if (_configuration["LIBRUM_SELFHOSTED"] == "true"){
+			var domain =_configuration["CleanUrl"];
+			var encodedToken=System.Web.HttpUtility.HtmlEncode(token);
+			resetLink = $"{domain}/user/resetPassword?email={user.Email}&token={encodedToken}";
```
Comment (DavidLazarescu) [inline src/Application/Utility/EmailSender.cs:63]: I don't understand what you are doing here. Could you give me an example?

**Review by DavidLazarescu — CHANGES_REQUESTED** [review]
This looks really good! I added some notes to your code, please look over them and change the PR respectively.

**Comment by 3036662** [comment]
Hi, thanks, I'll do everything on Monday.

File: src/Application/Managers/BookLocalStorageManager.cs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    private readonly string _bookCoverPrefix = "cover_";
+	private string dataDir;
+ 	private string booksDir;
+ 	private string coversDir;
+    private string baseDir;
+   
+    public BookLocalStorageManager()
+    {
+        baseDir=System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal);
+		dataDir= baseDir+"/librum_srorage";
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/BookLocalStorageManager.cs]: fixed

File: src/Application/Managers/BookLocalStorageManager.cs [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+using Application.Interfaces.Managers;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Net.Http.Headers;
+
+namespace Application.Managers;
+
+public class BookLocalStorageManager : IBookBlobStorageManager
+{
+    private readonly string _bookCoverPrefix = "cover_";
+	private string dataDir;
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/BookLocalStorageManager.cs:11]: fixed

File: src/Application/Managers/BookLocalStorageManager.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    {
+		var path=booksDir+"/"+guid;
+		 await Task.Run(() => System.IO.File.Delete(path));
+    }
+
+    public async Task<long> ChangeBookCover(Guid guid, MultipartReader reader)
+    {
+		  //if already exists
+		var filename=coversDir+"/"+guid;
+		if ( System.IO.File.Exists(filename)){
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/BookLocalStorageManager.cs:101]: fixed

File: src/Application/Managers/UserLocalStorageManager.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		if (!System.IO.File.Exists(filename)){
+			throw new CommonErrorException(400,"file not exists "+filename ,0);
+		}
+		return Task.FromResult<Stream>(File.OpenRead(filename));
+    }
+
+    public async Task ChangeProfilePicture(string guid, MultipartReader reader)
+    {
+        //if already exists
+		var filename=profilesDir+"/"+guid;
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/UserLocalStorageManager.cs:39]: fixed

File: src/Application/Managers/UserLocalStorageManager.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+{
+    private readonly string _profilePicturePrefix = "profilePicture_";
+    private string dataDir;
+    private string profilesDir;
+    private string baseDir;
+	
+    public UserLocalStorageManager()
+    {
+        baseDir=System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal); // $HOME folder 
+	    dataDir= baseDir+"/librum_srorage";
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/UserLocalStorageManager.cs:18]: fixed

File: src/Application/Managers/UserLocalStorageManager.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+using Application.Interfaces.Managers;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Net.Http.Headers;
+
+namespace Application.Managers;
+
+public class UserLocalStorageManager : IUserBlobStorageManager
+{
+    private readonly string _profilePicturePrefix = "profilePicture_";
+    private string dataDir;
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/UserLocalStorageManager.cs:11]: moved to local variable

File: src/Application/Managers/UserLocalStorageManager.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Managers/UserLocalStorageManager.cs:39]: fixed

File: src/Application/Services/AIService.cs [hunk 8]
```diff
@@ -34,6 +34,12 @@ public class AiService : IAiService
     public async Task ExplainAsync(string email, HttpContext context, string text, 
                                    string mode)
     {
+		//  return error mesage if no OpenAIToken exists for self-hosted mode, dont throw exception
+	    if (_configuration["LIBRUM_SELFHOSTED"] == "true" &&  String.IsNullOrEmpty(_configuration["OpenAIToken"])){
+            await context.Response.WriteAsync("data: " + "OpenAI temporary unavailable" + "\n");
```
[the remark this answers is not in the document]
Comment (3036662) [reply src/Application/Services/AIService.cs]: done

File: src/Application/Utility/EmailSender.cs (hunk 1 above)
Comment (3036662) [reply src/Application/Utility/EmailSender.cs:63]: For example:
1. self-hosted server is listening to https  127.0.0.1:5001 , 
2. it's exposed to network as mybook.com (standart port 443)
3. _configuration["CleanUrl"]  = https://mybooks.com, 
4. user will get email with password  reset link starting with mybooks.com/user/resetPassword.....  (not   127.0.0.1:5001).  
---------------
PS. its possible that server isn't aware about  its public url . Server running on 127.0.0.1:5001 and exposed to web by nginx.
So , the _configuration["CleanUrl"] is used to build "password reset" link with "public url".

Comment (DavidLazarescu) [reply src/Application/Utility/EmailSender.cs:63]: I understand. Thanks!

**Review by 3036662 — COMMENTED** [review]
Nice!

**Comment by DavidLazarescu** [comment]
Btw. were you able to run the self-hosted version on your machine? We still need to resolve all of the configuration lookup stuff like JWTKey, ... that are requested via e.g. `configuration["JWTKey"]`

**Comment by 3036662** [comment]
> Btw. were you able to run the self-hosted version on your machine? We still need to resolve all of the configuration lookup stuff like JWTKey, ... that are requested via e.g. `configuration["JWTKey"]`

Yes, I've tested everything , I'll give you full explanation in 30 min.

**Comment by 3036662** [comment]
> Btw. were you able to run the self-hosted version on your machine? We still need to resolve all of the configuration lookup stuff like JWTKey, ... that are requested via e.g. `configuration["JWTKey"]`

Sorry for delay. 
So, to run normally you need:
1. install dependencies
2. build with dotnet restore & dotnet build
3. Install server
-------------------------
normally to run server on linux you should run it as service, same does mysql and other servers, but it possible to run it simply with command line, first I'll talk about "classic" way:
1. create user for server running
2. create .service file for systemd
3. configure "secret variables" like "JWTKey" etc. - service secret configuration should be store in linux /etc folder in separate file, so ```systemctl show``` command will not show them.
4. create folder in "special user" home directory and install server
5. manage file permissions, to forbid other users to read from server folder and no write permissions, of course.
6. create ssl certificate
7. when everything is done, you can run it as usual service - systemctl start server-name
----------------------------------
i will push now self-hosted- build and install  instruction draft.
-----------------------------------
The way to run it from command line - for testing:
1. build and restore
2. create ssl certificate
3. configure server ports to listen
4. simply export all secret variables 
```export secretVariable=secretValue```
5. run with dotnet run

**Comment by DavidLazarescu** [comment]
Ok, I understand.

It would be good if the README.md of the server would contain a detailed section on this. If you could write the general outline of it, I could improve it and add examples, ...

**Comment by 3036662** [comment]
> Ok, I understand.
> 
> It would be good if the README.md of the server would contain a detailed section on this. If you could write the general outline of it, I could improve it and add examples, ...


I've added a manpage and self-hosted-install.md  drafts to new folder "self-hosting" , i've also added config files examples.
Please, take a look, so we can make them perfect.
I've tested the algorithm few times on different configuration, but still something may be missing for some linux versions.

**Comment by DavidLazarescu** [comment]
Looks great! I'll have more time in an hour or so, but I skimmed over it and it looks very detailed, which is great.

I really appreciate your work.

**Comment by DavidLazarescu** [comment]
In your guide:

`apt-get install dotnet-sdk-7.0 dotnet-7.0  dotnet-host openssl mariadb-server dotnet-7.0 dotnet-aspnetcore-runtime-7.0 openssl`
can be reduced to `sudo apt install dotnet-sdk-7.0 openssl mariadb-server`

and `install librum-server.service -m660  /etc/systemd/system/` gives me the error: `install: cannot stat 'librum-server.service': No such file or directory` on ubuntu

**Comment by 3036662** [comment]
> can be reduced to
it is not always absolutely reliable to entrust the resolution of dependencies to the package manager, but there should be no problems with this in ubuntu, but you can shorten it, he will pull up the necessary packages himself


> and `install librum-server.service -m660 /etc/systemd/system/`
``` install self-hosting/librum-server.service -m660  /etc/systemd/system/ ```
.service file is located in subfolder ```self-hosting```

**Comment by DavidLazarescu** [comment]
> .service file is located in subfolder `self-hosting` , try this, in case of problems, tell me, i will test install ubuntu on virtual environment and try to install.

Ah, I see. It seems like `PR_techSelection` contains the correct changes but `gitHubFork_selfHosted` contains the readme. I have cloned the repository with the readme which didn't have the selfhosted folder.

Is there a reason for having these two different branches, or can they be combined into one?

**Comment by 3036662** [comment]
> Is there a reason for having these two different branches, or can they be combined into one?
For now it looks like we don't need gitHubFork_selfHosted branche. I'll deal with it.

**Comment by DavidLazarescu** [comment]
Also, your step to create a new user fails with the following errors:
image

**Comment by 3036662** [comment]
> Also, your step to create a new user fails

Looks like you need to do it another way in ubuntu, try this
useradd -r -g librum-server -d /var/lib/librum-server --shell /usr/sbin/nologin

**Comment by DavidLazarescu** [comment]
And where exactly is `INSTALL.md` mentioned in `install -m664 INSTALL.md /var/lib/librum-server/srv`?

**Comment by 3036662** [comment]
> I think you could change it to something like: `sudo /usr/sbin/useradd -r -g librum-server -d /var/lib/librum-server -s /usr/sbin/nologin -c "Server for librum-reader" librumserver`

I'ts just linux-destribution - specific stuff, we can change everything to fit ubuntu, for me no diffirence, we already have rpm binary packages for our distro)
For now, i dont have an installed ubuntu, but i can test it with ubuntu if needed.

**Comment by 3036662** [comment]
> And where exactly is `INSTALL.md` mentioned in `install -m664 INSTALL.md /var/lib/librum-server/srv`?

fixed instruction, it was renamed to self-hosted-install.md

**Comment by DavidLazarescu** [comment]
> > I think you could change it to something like: `sudo /usr/sbin/useradd -r -g librum-server -d /var/lib/librum-server -s /usr/sbin/nologin -c "Server for librum-reader" librumserver`
> 
> I'ts just linux-destribution - specific stuff, we can change everything to fit ubuntu, for me no diffirence, we already have rpm binary packages for our distro) For now, i dont have an installed ubuntu, but i can test it with ubuntu if needed.

Yes, I think we should aim the instructions for ubuntu and provide specifics for other platforms when needed. I can do that later, for now I am just trying to run the self-hosted server on my ubuntu machine.

On ubuntu, `/etc/my.cnf.d/server.cnf` does not exist, there are `conf.d/          debian.cnf       debian-start     mariadb.cnf      mariadb.conf.d/  my.cnf           my.cnf.fallback ` in the `/etc/mysql` folder

**Comment by 3036662** [comment]
> my.cnf

Notmally mysql settings are located in my.cnf , but it can differ depending on version and distro and package version.
I'm already downloading ubuntu)

**Comment by DavidLazarescu** [comment]
> > my.cnf
> 
> Notmally mysql settings are located in my.cnf , but it can differ depending on version and distro and package version. I'm already downloading ubuntu)

I see, I found `/etc/mysql/mariadb.conf.d/50-server.cnf` which contains the `bind-address` which you mentioned. I'll continue with this file for now then.

**Comment by 3036662** [comment]
I,ve tested build&run process in Ubuntu 22.04, added instructions.

**Comment by DavidLazarescu** [comment]
Ok I have followed the instructions. I had some problems with `sudo mysql_secure_installation` but I changed my password and was able to complete it.

I think you miss spelled this: `vim /etc/librum-reader/librum-reader.conf`, you probably meant `vim /etc/librum-server/librum-server.conf`

**Comment by DavidLazarescu** [comment]
Also, don't you need to add `export LIBRUM_SELFHOSTED=true` to the run.sh script?

**Comment by DavidLazarescu** [comment]
And how exactly is the server supposed to find the config values in `vim /etc/librum-server/librum-server.conf`? Starting the application manually fails for me since it can't find the JWT token.

**Comment by 3036662** [comment]
> I think you miss spelled this: `vim /etc/librum-reader/librum-reader.conf`

Yes, thanks. Fixed.

**Comment by 3036662** [comment]
> Also, don't you need to add `export LIBRUM_SELFHOSTED=true` to the run.sh script?

No, we don't need to export it is exported when systemd will run service ( you can look into librum-server.service file, its exported there)

**Comment by 3036662** [comment]
> how exactly is the server supposed to find the config value

Secret values will be loaded with systemd when starting service.  You can find 

``` EnvironmentFile=/etc/librum-server/librum-server.conf ``` 

in ```librum-server.service``` file

**Comment by DavidLazarescu** [comment]
> > Also, don't you need to add `export LIBRUM_SELFHOSTED=true` to the run.sh script?
> 
> 
> 
> No, we don't need to export it is exported when systemd will run service ( you can look into librum-server.service file, its exported there)

Ah, I see now. It probably didn't work because of the wrong file name then.

**Comment by DavidLazarescu** [comment]
Thanks for the elaboration. I missed the line with `librum-server.conf` in the service file.

**Comment by DavidLazarescu** [comment]
Ok, it now finds the configuration, but I am having problems with my mariadb-server installation.

How exactly would I be supposed to continue here?
image

(I have previously found a solution where I reset the password and then was able to login, but it seems to have messed up something so that the service wasnt starting anymore. I have no purged everything and reinstalled it.)

**Comment by 3036662** [comment]
> Thanks for the elaboration. I missed the line with `librum-server.conf` in the service file.

It's possible to export all variables in .service file, but we have "secret" variables, so it's better to have them in separate file with no read-write permission for other users.

**Comment by DavidLazarescu** [comment]
And some other small things:
- `sudo apt-get mariadb-server` doesn't work, you are missing the `install` keyword. I think we should generally use `sudo apt install ...` everywhere since this is the newer variant
- "go to /var/lib/librum-server/srv/ and add to appsettings.json and appsettings.Development.json - this block:" do we really need to add it to the `appsettings.Development.json`? We are providing instructions for a release build, so afaik. it should never be read anyways

**Comment by 3036662** [comment]
> How exactly would I be supposed to continue here?

If your have new installation of mariadb - there is no root password, normally you simply press enter key, then you will be asked for new password for root.

**Comment by DavidLazarescu** [comment]
And which database &Uid should I use in the connection string?
image

**Comment by 3036662** [comment]
> * We are providing instructions for a release build

right, it's not necessary

> * sudo apt install
ok, it will work too.



> * sudo apt-get mariadb-server

fixed

**Comment by 3036662** [comment]
> And which database

You can write any valid db name directly to connection string, it will be created by server.

**Comment by DavidLazarescu** [comment]
> > And which database
> 
> You can write any valid db name directly to connection string, it will be created by server.

Great, it works. I am able to call the API.

**Comment by DavidLazarescu** [comment]
Btw, I think it might be nice to remove all of the email related stuff, such as confirming the account via a confirmation email. It is imo. annoying to need to setup SMTP credentials for this stuff.

**Comment by 3036662** [comment]
> nice to remove

It's always easy way to delete some functionality, but not very easy to add it. 
I tried not to reduce the possibilities but to increase them.

For home use you can just push 1 value to whole database column (AspNetUsers->EmailConfirmed) all users will become verified. In my opinion - "delete unconfirmed users it's very good thing. I've created a dozen users while testing, and all of them were automatically deleted by server because of unconfirmed email.

**Comment by DavidLazarescu** [comment]
> > nice to remove
> 
> It's always easy way to delete some functionality, but not very easy to add it. I tried not to reduce the possibilities but to increase them.
> 
> For home use you can just push 1 value to whole database column (AspNetUsers->EmailConfirmed) all users will become verified. In my opinion - "delete unconfirmed users it's very good thing. I've created a dozen users while testing, and all of them were automatically deleted by server because of unconfirmed email.

Hm, I get your point, but I don't think that its worth the extra struggle for the end user. I think by far most people who want to self-host Librum will simply setup one account and be happy with it. If they want an account deleted, they can simply delete it by logging into it and clicking "Delete Account" in the settings.

Setting up an extra SMTP service to simply use Librum is a big burden imo. so keeping it as simple as possible would be good.

**Comment by 3036662** [comment]
> Setting up an extra SMTP service to simply use Librum is a big burden imo. so keeping it as simple as possible would be good.

I Never can argue with "Keep it simple" ! )

**Comment by DavidLazarescu** [comment]
From what I see the server only sends emails in two places: When registering and when resetting the password.

I have already removed the email sending while registering. I'll now look into the password resetting.

**Comment by DavidLazarescu** [comment]
The only option to avoid the email sending in the password reset is to open a browser with the form right from the server, but this is not a very nice way to handle this.
Do you have another idea?

**Comment by 3036662** [comment]
I believe that a user who can install and configure the database and server service in linux will not be confused by the mail setup, its basics, no rocket since.
Still, the ability to reset the password and a login  tied to the mail is not bad, you can run a server, for example, for a whole house, for neighbours.
But, I won't argue.
If we continue with disabling simple checks, it will transform to simple local storage. I would consider to save full functionality.

**Comment by DavidLazarescu** [comment]
> I believe that a user who can install and configure the database and server service in linux will not be confused by the mail setup, its basics, no rocket since. Still, the ability to reset the password and a login tied to the mail is not bad, you can run a server, for example, for a whole house, for neighbours. But, I won't argue. If we continue with disabling simple checks, it will transform to simple local storage. I would consider to save full functionality.
> 
> IMHO, the program is primarily interesting for the possibility of launching a full-fledged server.
> 
> If we disable sending email when password reset, we can same way disable password, i think. It's about local storage reader, or about server without password, i don't sure it's worth "super simple installation".

You are right, I did not think about the fact that not sending an email to the account's owner on a password reset would basically make passwords useless and thus destroy all of the security.

I will revert the last commit I made and we will keep the email sending for self-hosted servers.

**Comment by DavidLazarescu** [comment]
Ok, it is running just fine together with the client application.
I'll do some more testing tomorrow and if everything goes well we can merge this soon.

**Review by DavidLazarescu — APPROVED** [review]

**Comment by 3036662** [comment]
It looks like you're putting things in order today. good job!

**Comment by 3036662** [comment]
> Ok, everything looks good so far. Have you tested if password resetting works fine for the self-hosted server? I don't have a self-hosted email, thus I can not test it.
> 
> Can you confirm that it works fine?

You can test it with any public email service like gmail, all, etc. I' be made some test, it works fine with my email server.

**Comment by 3036662** [comment]
> > Ok, everything looks good so far. Have you tested if password resetting works fine for the self-hosted server? I don't have a self-hosted email, thus I can not test it.
> > Can you confirm that it works fine?
> 
> You can test it with any public email service like gmail, aol, etc. I' ve made some tests, it works fine with my email server.

I mean you don't need self-hosted email. You can use any public email, if it allows smtp mail client usage.

**Comment by DavidLazarescu** [comment]
> It looks like you're putting things in order today. good job!

Yes, pretty much everything should be done now. Could you change the PR request to merge into dev/develop instead of main though?

**Comment by 3036662** [comment]
> > It looks like you're putting things in order today. good job!
> 
> Yes, pretty much everything should be done now. Could you change the PR request to merge into dev/develop instead of main though?

Done.

**Pull request merged**